src/audio/module_adapter/module_adapter.c

```diff
@@ -879,7 +879,11 @@ static inline int module_adapter_audio_stream_type_copy(struct comp_dev *dev)
 
 	/* handle special case of HOST/DAI type components */
 	if (dev->ipc_config.type == SOF_COMP_HOST || dev->ipc_config.type == SOF_COMP_DAI)
+#if CONFIG_IPC_MAJOR_3
 		return module_process_legacy(mod, NULL, 0, NULL, 0);
+#else
+		return mod->priv.ops->process_audio_stream(mod, NULL, 0, NULL, 0);
```

Comment:
This should be 0 cost with the staic inline in the header - I know we do check the ops != NULL, so that will add cost, but we shouldcheck all mandarory ops != NULL at init and check all optional ops != NULL at every usage.

Reply:
This can be optimised, the md->ops will be known at init time, we dont need to check this at process time to know which ops function to call.

The return value check can also be improved by only checking (ret < 0), we should not be filtering error numbers. Likewise modules should only return -ENODATA or -ENOSPC iff these are real errors that stop the module from processing any further data (without a reset).

```c
int module_process_legacy(struct processing_module *mod,
			  struct input_stream_buffer *input_buffers, int num_input_buffers,
			  struct output_stream_buffer *output_buffers,
			  int num_output_buffers)
{
	struct comp_dev *dev = mod->dev;
	int ret;

	struct module_data *md = &mod->priv;

	comp_dbg(dev, "module_process_legacy() start");

#if CONFIG_IPC_MAJOR_3
	if (md->state != MODULE_IDLE) {
		comp_err(dev, "module_process(): wrong state of comp_id %x, state %d",
			 dev_comp_id(dev), md->state);
		return -EPERM;
	}

	/* set state to processing */
	md->state = MODULE_PROCESSING;
#endif
	if (md->ops->process_audio_stream)
		ret = md->ops->process_audio_stream(mod, input_buffers, num_input_buffers,
						    output_buffers, num_output_buffers);
	else if (md->ops->process_raw_data)
		ret = md->ops->process_raw_data(mod, input_buffers, num_input_buffers,
						output_buffers, num_output_buffers);
	else
		ret = -EOPNOTSUPP;

	if (ret && ret != -ENOSPC && ret != -ENODATA) {
		comp_err(dev, "module_process() error %d: for comp %d",
			 ret, dev_comp_id(dev));
		return ret;
	}

	comp_dbg(dev, "module_process_legacy() done");

#if CONFIG_IPC_MAJOR_3
	/* reset state to idle */
	md->state = MODULE_IDLE;
#endif
	return 0;
}
```

and for

```c
int module_adapter_copy(struct comp_dev *dev)
{
	comp_dbg(dev, "module_adapter_copy(): start");

	struct processing_module *mod = comp_get_drvdata(dev);

	if (IS_PROCESSING_MODE_AUDIO_STREAM(mod))
		return module_adapter_audio_stream_type_copy(dev);

	if (IS_PROCESSING_MODE_RAW_DATA(mod))
		return module_adapter_raw_data_type_copy(dev);

	if (IS_PROCESSING_MODE_SINK_SOURCE(mod))
		return module_adapter_sink_source_copy(dev);

	comp_err(dev, "module_adapter_copy(): unknown processing_data_type");
	return -EINVAL;
}
```
This can be improved by moving the mode check to init time rather than process time.

Reply:
module_init already checked:

	/*check interface, there must be one and only one of processing procedure */
	if (!interface->init || !interface->prepare ||
	    !interface->reset || !interface->free ||
	    (!!interface->process + !!interface->process_audio_stream +
	     !!interface->process_raw_data != 1)) {
		comp_err(dev, "module_init(): comp %d is missing mandatory interfaces",
			 dev_comp_id(dev));
		return -EIO;
	}

However, we still need classify which process is valid during each processing:
process
process_raw_data
process_audio_stream

to avoid IS_PROCESSING_MODE_AUDIO_STREAM(mod), I need add a flag in mod to indicate which process is valid.
then use switch: module_adapter_copy, like below:

switch (mod->process_module) {
case 1:
   process_audio_stream;
case 2:
   process_raw_data;
case 3: process;
}

something like this, any more good suggestions?

do you think it is not allowed or don't want to add more #if IPC3?
Later, I want to submit a PR to separate module_adapter.c to ipc3/ipc4, then inside #if will be totally disappear.

Reply:
Since we check for ops != NULL in init then I would expect

```c

// header
static inline int module_process(mod, blah, blah) {
     return mod->priv.ops->process_audio_stream(mod, NULL, 0, NULL, 0);
}

// C file
#if CONFIG_IPC_MAJOR_3
		return module_process_legacy(mod, NULL, 0, NULL, 0);
#else
		return module_process(mod, NULL, 0, NULL, 0);
#endif
```